Support for linear gradients: add additional type to waveColor and progressColor parameters

califken opened the pull request:
### Short description of changes:
Added feature to pass an array of colors into the waveColor and progressColor parameters.  Edited the setFillStyles function to check if the provided argument is a string or an object.  In the case of an object, a linear gradient is created, with each item in the array added as a color stop.

![wavecolorgradient](https://user-images.githubusercontent.com/2645777/130211579-9347aa5a-ca9d-48d9-9e03-110b2953d610.png)
![wavecolorgradientcodeexample](https://user-images.githubusercontent.com/2645777/130211583-e6d5f064-5f14-48b7-ab36-fc27d51642b0.png)

### Breaking in the external API:
Not sure.

### Breaking changes in the internal API:
None, as it runs the check to see if it is a string, therefore should be backwards compatible.

### Todos/Notes:
If this feature is integrated, I would eventually like to integrate the ability to pass images and even a video as the wave background.

### Related Issues and other PRs:

thijstriemstra reviewed (changes requested):
That's great @califken! Can you also add an example that showcases this feature? And it'll also need a changelog entry? And make sure to run `npm run lint` because there's a lint error: https://github.com/katspaugh/wavesurfer.js/pull/2345/checks?check_run_id=3384740565#step:8:26

califken commented:
> That's great @califken!
> 
> * Can you also add an example that showcases this feature?
Done, added a gradient-fill-styles directory in the examples folder.  Not sure if there is an index that I should add it to somewhere.
> * And it'll also need a changelog entry?
Done, added a changelog entry to CHANGES.md.  Bumped the version up to 5.3.0.
> * And make sure to run `npm run lint` because there's a lint error: https://github.com/katspaugh/wavesurfer.js/pull/2345/checks?check_run_id=3384740565#step:8:26
Corrected the issue.
> * add a unit test
Added @test {WaveSurfer#setWaveColorGradient} to wavesurfer.spec.js
> Also, what does it look like with bars (#1722)?

It looks awesome with bars!  :)
![Screen Shot 2021-08-20 at 3 41 35 PM](https://user-images.githubusercontent.com/2645777/130300237-262d685e-5451-4722-8717-31b09c0d56df.png)

califken commented:
Rebased and pushed the one commit.  Ready to go.

califken commented on CHANGES.md:
```diff
@@ -1,5 +1,8 @@
 wavesurfer.js changelog
 =======================
+5.3.0 (20.08.2021)
```
changed to unreleased

califken commented on example/gradient-fill-styles/main.js:
```diff
@@ -0,0 +1,81 @@
+    button.addEventListener('click', wavesurfer.playPause.bind(wavesurfer));
+
+
+    // WaveSurfer with bars example
```
updated comment

califken commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "wavesurfer.js",
-  "version": "5.2.0",
+  "version": "5.3.0",
```
changed back to 5.2.0

califken commented on example/gradient-fill-styles/index.html:
```diff
@@ -0,0 +1,180 @@
+
+<head>
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+    <title>wavesurfer.js | Waveform using bars</title>
```
changed title to wavesurfer.js | Gradient Fill Styles Example

thijstriemstra commented on src/wavesurfer.js:
```diff
@@ -1154,8 +1155,9 @@ export default class WaveSurfer extends util.Observer {
-     * @param {string} color A CSS color string.
+     * @param {string|object} color A CSS color string, or an array of CSS color strings.
      * @example wavesurfer.setProgressColor('#400');
+     * @example wavesurfer.setWavecolor(['blue','black','yellow']);
```
setWaveColor example doesn't belong here.

califken replied:
Agreed, removed in latest commit.

msuhov commented on src/drawer.canvasentry.js:
```diff
@@ -146,14 +146,25 @@ export default class CanvasEntry {
+        }
         if (this.hasProgressCanvas) {
-            this.progressCtx.fillStyle = progressColor;
+            if (typeof progressColor == 'string') {
```
it looks like a duplicate of previous code, I'd rather extract it to an utility or something like this:
```
this.waveCtx.fillStyle = getFillStyle(this.waveCtx, waveColor)

if (this.hasProgressCanvas) {
  this.progressCtx.fillStyle = getFillStyle(this.progressCtx, progressColor)
}


...

export const getFillStyle = (context, color) => {
  if (typeof waveColor == 'string') {
    return color
  }

  const waveGradient = context.createLinearGradient(0, 0, 0, context.canvas.height);
  color.forEach((value, index) => waveGradient.addColorStop((index / color.length), value));

  return waveGradient
}
```

califken replied:
Agreed, included in latest commit.

thijstriemstra commented on spec/wavesurfer.spec.js:
```diff
@@ -264,6 +264,27 @@ describe('WaveSurfer/playback:', function() {
+        wavesurfer.setWaveColor(colors);
+        const waveColor = wavesurfer.getWaveColor();
+
+        expect(waveColor).toEqual([
```
`toEqual(colors)`?

thijstriemstra commented on src/drawer.canvasentry.js:
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
      *
-     * @param {string} waveColor Fill color for the wave canvas
-     * @param {?string} progressColor Fill color for the progress canvas
+     * @param {string|string[]} waveColor Fill color for the wave canvas, or an array of colors to apply as a gradient
```
spread sentence over 2 lines

thijstriemstra commented on src/drawer.canvasentry.js:
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
-     * @param {string} waveColor Fill color for the wave canvas
-     * @param {?string} progressColor Fill color for the progress canvas
+     * @param {string|string[]} waveColor Fill color for the wave canvas, or an array of colors to apply as a gradient
+     * @param {?string|string[]} progressColor Fill color for the progress canvas, or an array of colors to apply as a gradient
```
spread sentence over 2 lines

thijstriemstra commented on src/drawer.canvasentry.js:
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
+     *
+     * @param {CanvasRenderingContext2D} ctx Rendering context of target canvas
+     * @param {string|string[]} color Fill color for the wave canvas, or an array of colors to apply as a gradient
+     * @returns {string | CanvasGradient} Returns a string fillstyle value, or a canvas gradient
```
add version marker for 5.3.0 to method (https://jsdoc.app/tags-since.html)

thijstriemstra replied:
`@returns {string | CanvasGradient}`: remove extra spaces

thijstriemstra reviewed (changes requested):
Build is also failing.

thijstriemstra reviewed (approved):
thx!

jadenmazzone commented:
@thijstriemstra is it possible to have the linear gradient show left to right - photo attached to see what I mean
<img width="891" alt="Screen Shot 2022-12-07 at 11 20 19 AM" src="https://user-images.githubusercontent.com/19211154/206275659-59091e04-44ec-4ed4-a001-0c967de503e3.png">
